src/ui/sign_tx_nbgl.c

```diff
@@ -0,0 +1,284 @@
+    static_items[static_items_nb].value = G_tx.total_fees;
+    ++static_items_nb;
+
+    static_items[static_items_nb].item = "Valid until height";
+    static_items[static_items_nb].value = G_tx.valid_until_block;
+    ++static_items_nb;
+
+    for (int i = 0; i < G_context.tx_info.transaction.signers_size; ++i) {
+        dyn_items[dyn_items_nb].kind = SIGNER;
+        dyn_items[dyn_items_nb].content.as_item_signer.signer_index = i;
+        ++dyn_items_nb;
+        dyn_items[dyn_items_nb].kind = ACCOUNT;
+        dyn_items[dyn_items_nb].content.as_item_account.signer_index = i;
+        ++dyn_items_nb;
+        dyn_items[dyn_items_nb].kind = SCOPE;
+        dyn_items[dyn_items_nb].content.as_item_scope.signer_index = i;
+        ++dyn_items_nb;
+
+        signer_t signer = G_context.tx_info.transaction.signers[i];
+        for (int j = 0; j < signer.allowed_contracts_size; ++j) {
+            dyn_items[dyn_items_nb].kind = CONTRACT;
+            dyn_items[dyn_items_nb].content.as_item_contract.signer_index = i;
+            dyn_items[dyn_items_nb].content.as_item_contract.contract_index = j;
+            ++dyn_items_nb;
+        }
+        for (int j = 0; j < signer.allowed_groups_size; ++j) {
+            dyn_items[dyn_items_nb].kind = GROUP;
+            dyn_items[dyn_items_nb].content.as_item_group.signer_index = i;
+            dyn_items[dyn_items_nb].content.as_item_group.group_index = j;
+            ++dyn_items_nb;
+        }
+    }
+
+    return 0;
+}
+
+static const nbgl_pageInfoLongPress_t review_final_long_press = {
+    .text = "Sign message on\nNeo N3 network?",
+    .icon = &C_icon_neo_n3_64x64,
+    .longPressText = "Hold to sign",
+    .longPressToken = 0,
+    .tuneId = TUNE_TAP_CASUAL,
+};
+
+static void review_final_callback(bool confirmed) {
+    if (confirmed) {
+        ui_action_validate_transaction(true, false);
+        nbgl_useCaseStatus("MESSAGE\nSIGNED", true, ui_menu_main);
+    } else {
+        rejectUseCaseChoice();
+    }
+}
+
+static void rejectChoice(void) {
+    ui_action_validate_transaction(false, false);
+    nbgl_useCaseStatus("message\nrejected", false, ui_menu_main);
+}
+
+static void rejectUseCaseChoice(void) {
+    nbgl_useCaseConfirm("Reject message?", NULL, "Yes, reject", "Go back to message", rejectChoice);
+}
+
+// function called by NBGL to get the current_pair indexed by "index"
+static nbgl_layoutTagValue_t *get_single_action_review_pair(uint8_t index) {
+    if (index < static_items_nb) {
+        // No need to copy to dyn_slots as item and value are pointers to static values
+        current_pair.item = static_items[index].item;
+        current_pair.value = static_items[index].value;
+    } else {
+        dynamic_item_t *current_item = &dyn_items[index - static_items_nb];
+        signer_t s = G_context.tx_info.transaction.signers[current_item->content.as_item_scope.signer_index];
+        dynamic_slot_t *slot = &dyn_slots[index % 4];
```

Comment:
Can you define a macro for the size of `dyn_slots` and replace the literal `4` here ?

Reply:
Done